backend/main/tests/services/matching/test_k8s_job_runner.py

```diff
@@ -0,0 +1,346 @@
+            self.addCleanup(patcher.stop)
+
+        self.mock_run_job = patches["_run_job"]
+        self.mock_wait_for_pod = patches["_wait_for_pod"]
+        self.mock_get_pod_logs = patches["_get_pod_logs"]
+        self.mock_stream_logs = patches["_stream_pod_logs_until_job_completion"]
+        self.mock_get_state = patches["_get_pod_container_state"]
+
+    def test_run_job_completed(self) -> None:
+        """Method run_job should retrieve logs if the pod has already completed."""
+        self.mock_k8s.wait_for_job_pods.return_value = [
+            PodState(name="pod-1", phase="Succeeded")
+        ]
+        self.mock_k8s.get_pod_logs.return_value = "line1\nline2"
+        self.mock_k8s.get_pod_container_states.return_value = [
+            ContainerState(
+                waiting=None,
+                terminated=ContainerTerminatedState(
+                    exit_code=0,
+                    finished_at="",
+                    reason="Completed",
+                    started_at="",
+                ),
+            )
+        ]
+
+        result = self.runner.run_job()
+
+        self.assertEqual(result, JobResult(return_code=0, error_message=None))
+
+        self.mock_run_job.assert_called_once_with("matcher-job")
+        self.mock_wait_for_pod.assert_called_once_with("matcher-job")
+        self.mock_get_pod_logs.assert_called_once_with("pod-1")
+        self.mock_stream_logs.assert_not_called()
+        self.mock_get_state.assert_called_once_with("pod-1")
+        self.mock_k8s.delete_job.assert_called_once_with("matcher-job")
+        self.mock_k8s.wait_for_job_deletion.assert_called_once_with("matcher-job")
+
+    def test_run_job_running(self) -> None:
+        """Method run_job should stream logs if pod is running."""
+        self.mock_k8s.wait_for_job_pods.return_value = [
+            PodState(name="pod-1", phase="Running")
+        ]
+        self.mock_k8s.stream_pod_logs.return_value = ["log-1", "log-2"]
+        self.mock_k8s.get_pod_container_states.return_value = [
+            ContainerState(
+                waiting=None,
+                terminated=ContainerTerminatedState(
+                    exit_code=0,
+                    finished_at="",
+                    reason="Completed",
+                    started_at="",
+                ),
+            )
+        ]
+
+        result = self.runner.run_job()
+
+        self.assertEqual(result, JobResult(return_code=0, error_message=None))
+
+        self.mock_run_job.assert_called_once_with("matcher-job")
+        self.mock_wait_for_pod.assert_called_once_with("matcher-job")
+        self.mock_get_pod_logs.assert_not_called()
+        self.mock_stream_logs.assert_called_once_with("matcher-job", "pod-1")
+        self.mock_get_state.assert_called_once_with("pod-1")
+        self.mock_k8s.delete_job.assert_called_once_with("matcher-job")
+        self.mock_k8s.wait_for_job_deletion.assert_called_once_with("matcher-job")
+
+    def test_run_job_failure(self) -> None:
```

Comment:
We can also add job already completed with failure - right now we only test for "Running" job that fails.